### PR TITLE
add resize_token_embedding functionality

### PR DESCRIPTION
**Pull request by iamzainhuda**
This PR introduces a resize_token_embeddings method which allows users to add additional tokens to be trained as part of fine tuning. It supports both single and multi device setups. The function can be easily extended to support additional initialization strategies.

For the distributed case, we unshard() the token embedding and output layer before passing into resize_token_embedding(). This way we can use the same logic as the single device case as the parameters are gathered on to the rank. Once complete, we fully_shard() the new token embedding and output layers utilizing the same mesh used for the initial sharding. This is the recommended way to resize FSDP2 sharded parameters. Thanks to @weifengpy for his guidance on the FSDP2 sharding mechanism.

The usage is,

On single device and distributed setups:

```python
model = setup_model(...)
tokenizer = setup_tokenizer(...)
resize_token_embeddings(model, tokenizer.vocab_size)
```

When using a recipe, (right now enabled for `full_finetune_distributed` and `full_finetune_single_device`), add `resize_token_embeddings = True` in YAML.
Example,
```yaml
...
max_steps_per_epoch: null
gradient_accumulation_steps: 1  # Use to increase effective batch size
clip_grad_norm: null
compile: False  # torch.compile the model + loss, True increases speed + decreases memory
optimizer_in_bwd: False  # True saves memory. Requires gradient_accumulation_steps=1
resize_token_embeddings: True
...
````

#### Follow up tasks:
- additional embedding weight init methods
- loading checkpoint from resized model requires model to be created with new vocab size
- FusionEmbedding resizing as part of TransformerDecoder
 
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
* Added resize_token_embedding() utility that allows users to flexibly change the size of the token embeddings and corresponding output projection.

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [x] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [x] run recipe tests via `pytest tests -m integration_test`
- [x] manually run any new or modified recipes with sufficient proof of correctness
- [x] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

Unit Testing:

Unit tests cover both single device and distributed cases, we check the new nn.Embedding size as well as the mean values of the new tokens to ensure they were init correctly. The E2E case covers testing a training run. We can see training completes successfully in the resized model that is sharded across 8 devices.

Without resizing, we can see special token init values on LLama 3 8B base model are very small, taking a look at token `<|start_header_id|> = 128006`
image

With resizing, we can see special tokens are init to the mean of 128k tokens on LLama 8B base:
image

##### E2E Test

Resizing the last 256 tokens in Llama 3 tokenizer to mean embedding of previous 120k tokens. Added an additional 4 tokens to resize the entire layer. We don't see significant difference in loss (run for 2 epochs, 20 steps each), but we show with the resized embedding in distributed setup we're able to finish training with at least equivalent loss.

image

vs. baseline
image

Running,
```
tune run --nproc_per_node 8 full_finetune_distributed --config recipes/configs/llama3/8B_full.yaml max_steps_per_epoch=20 epochs=2 checkpointer.checkpoint_dir=/tmp/Meta-Llama-3-8B
 ```

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [x] I have added an example to docs or docstrings

### CODE REVIEW

File: recipes/full_finetune_distributed.py [hunk 1]
```diff
@@ -340,6 +343,25 @@ def setup(self, cfg: DictConfig) -> None:
         )
         self._tokenizer = config.instantiate(cfg.tokenizer)
 
+        if self._tokenizer.vocab_size != self._model.tok_embeddings.num_embeddings:
+            # Unshard the FSDP wrapped token embeddings and output projection
+            # tok_embeddings and output are wrapped by fully_shard, we can call .unshard()
+            # direclty on those layers only
+            self._model.tok_embeddings.unshard()
```
Comment (ebsmothers) [inline recipes/full_finetune_distributed.py:350]: It's possible that `self._model` is not an instance of `TransformerDecoder`. E.g. for multimodal models we provide both [EarlyFusionModel](https://github.com/pytorch/torchtune/blob/5417a29ba96be78f11b23a67657ab4c4530b7f78/torchtune/modules/model_fusion/_early_fusion.py#L16) and [DeepFusionModel](https://github.com/pytorch/torchtune/blob/5417a29ba96be78f11b23a67657ab4c4530b7f78/torchtune/modules/model_fusion/_deep_fusion.py#L20), so we should at least be able to handle both of these gracefully. One option is to just do all this directly in the `resize_token_embeddings` function, otherwise we start to bloat the recipe code with lots of if/else checks.

File: recipes/full_finetune_distributed.py [hunk 2]
```diff
@@ -340,6 +343,25 @@ def setup(self, cfg: DictConfig) -> None:
         )
         self._tokenizer = config.instantiate(cfg.tokenizer)
 
+        if self._tokenizer.vocab_size != self._model.tok_embeddings.num_embeddings:
+            # Unshard the FSDP wrapped token embeddings and output projection
+            # tok_embeddings and output are wrapped by fully_shard, we can call .unshard()
+            # direclty on those layers only
```
Comment (ebsmothers) [inline recipes/full_finetune_distributed.py:349]: super nit
```suggestion
            # directly on those layers only
```

File: torchtune/modules/transforms/tokenizers/_utils.py [hunk 3]
```diff
@@ -12,7 +12,7 @@
 from torchtune.data._utils import truncate
 from torchtune.utils import get_logger
 
-_log: logging.Logger = get_logger()
+log: logging.Logger = get_logger()
```
Comment (ebsmothers) [inline torchtune/modules/transforms/tokenizers/_utils.py:15]: Can also probably revert the changes to this file and transformer.py now to match main

File: torchtune/modules/common_utils.py [hunk 4]
```diff
@@ -442,3 +447,95 @@ def delete_kv_caches(model: nn.Module):
         if hasattr(module, "kv_cache") and callable(module.kv_cache):
             module.cache_enabled = False
             module.kv_cache = None
+
+
+def resize_token_embeddings(model: TransformerDecoder, new_num_tokens: int) -> None:
+    """
+    Resizes the token embeddings and the final output projection layer of a TransformerDecoder model.
+
+    This function modifies the model in-place.
```
Comment (ebsmothers) [inline torchtune/modules/common_utils.py:456]: Would also explain (somewhere) in the docstring that you're initializing new positions using the mean of the old embeddings

File: torchtune/modules/common_utils.py [hunk 5]
```diff
@@ -442,3 +447,95 @@ def delete_kv_caches(model: nn.Module):
...
+    if new_num_tokens > old_num_tokens:
+        mean_embedding = new_embeddings.weight.data[:n].mean(dim=0)
+        new_embeddings.weight.data[n:] = mean_embedding.unsqueeze(0)
+
+    new_embeddings.weight.requires_grad_(old_embeddings.weight.requires_grad)
+    model.tok_embeddings = new_embeddings
+
+    output_layer = model.output
+    if isinstance(output_layer, TiedLinear):
+        if output_layer.tied_module is old_embeddings:
```
Comment (ebsmothers) [inline torchtune/modules/common_utils.py:512]: Dumb q: is this just checking equality of the underlying data_ptr?

**Review by ebsmothers — COMMENTED** [review]
Thanks for adding this and for the thorough testing! Left a bunch of comments but this is looking pretty good. Lmk if anything I suggested is unclear

File: recipes/full_finetune_distributed.py [hunk 6]
```diff
@@ -340,6 +343,25 @@ def setup(self, cfg: DictConfig) -> None:
...
+                self._model.output.unshard()
+
+            log.info(
+                f"Tokenizer vocab size ({self._tokenizer.vocab_size}) differs from model "
+                f"embedding size ({self._model.tok_embeddings.num_embeddings}). Resizing model embeddings..."
+            )
+            resize_token_embeddings(self._model, self._tokenizer.vocab_size)
+
+            # Reshard with new values, use the same mesh that was used in initial sharding
+            fully_shard(self._model.tok_embeddings, mesh=self.world_mesh[("dp_shard",)])
```
[the remark this answers is not in the document]
Comment (iamzainhuda) [reply recipes/full_finetune_distributed.py:361]: Yeah agreed, much cleaner to handle it within function vs. bloating all the recipes/callsites with additional setup just to use function - I'll make the change to infer from DTensor mesh.

On your idea, I agree that type of utility would be very nice to have (almost akin to summon_full_params from FSDP1, https://docs.pytorch.org/docs/stable/fsdp.html#torch.distributed.fsdp.FullyShardedDataParallel.summon_full_params). I imagine this use case would be quite common

File: recipes/full_finetune_distributed.py (hunk 1 above)
Comment (iamzainhuda) [reply recipes/full_finetune_distributed.py:350]: Got it - I'll move all of this within the function and handle for the various types of model's that can be passed in

File: torchtune/modules/common_utils.py (hunk 4 above)
Comment (iamzainhuda) [reply torchtune/modules/common_utils.py:456]: Added - I've also added "init_strategy" optional arg to the function. Abstracted the resize op to it's own helper as we add additional init strategies it will be cleaner to keep their implementations in a helper. I am thinking we can add this in a follow up PR and also allow for user's to pass in a callable for init strategy?

File: torchtune/modules/common_utils.py (hunk 5 above)
Comment (iamzainhuda) [reply torchtune/modules/common_utils.py:512]: this is checking if object points to the same memory as tok_embedding's nn.Embedding, underlying data_ptr() isn't checked. it is better check is to use data_ptr() instead if we have output and tok_embeddings as seperate nn.Embeddings for any case.

however, if output_layer is tied, we should always replace the tied_module with new tok_embedding - i feel like this check is not needed?

File: recipes/full_finetune_distributed.py [hunk 7]
```diff
@@ -319,6 +322,24 @@ def setup(self, cfg: DictConfig) -> None:
             ac_option=cfg.get("ac_option", None),
         )
         self._tokenizer = config.instantiate(cfg.tokenizer)
+       
+        if self._tokenizer.vocab_size != self._model.tok_embeddings.num_embeddings:
+            # Unshard the FSDP wrapped token embeddings and output projection
+            # tok_embeddings and output are wrapped by fully_shard, we can call .unshard()
+            # direclty on those layers only
+            self._model.tok_embeddings.unshard()
+            if not isinstance(self._model.output, TiedLinear): self._model.output.unshard()
```
[the remark this answers is not in the document]
Comment (iamzainhuda) [reply recipes/full_finetune_distributed.py]: If I understand correctly, the context manager would essentially handle the loop of d_tensor.full_tensor() --> user op --> redistribute_tensor(tensor, mesh)?

[the remark this answers is not in the document]
Comment (ebsmothers) [reply recipes/full_finetune_distributed.py]: Yeah exactly. Doesn't have to be done in this PR anyways, just an idea for a future quality-of-life improvement

File: torchtune/modules/common_utils.py [hunk 8]
```diff
@@ -442,3 +448,131 @@ def delete_kv_caches(model: nn.Module):
         if hasattr(module, "kv_cache") and callable(module.kv_cache):
             module.cache_enabled = False
             module.kv_cache = None
+
+
+def resize_token_embeddings(model: nn.Module, num_embeddings: int) -> None:
```
Comment (pbontrager) [inline torchtune/modules/common_utils.py:453]: This currently wouldn't work with fusion models or models that use the [FusionEmbedding](https://docs.pytorch.org/torchtune/stable/generated/torchtune.modules.model_fusion.FusionEmbedding.html#torchtune.modules.model_fusion.FusionEmbedding) layer. For a fusion model you want to check for the "decoder" and grab that instead of model.

To support resizing FusionEmbeddings you would want resize the embedding parameter and num_embeddings. It would also be reasonable to throw an error and not support resizing FusionEmbeddings as they're already a different technique for resizing embeddings but I don't think it would be too much harder to support them.

Comment (ebsmothers) [reply torchtune/modules/common_utils.py:453]: Ah yeah this is my fault. I was thinking that the usage of `FusionEmbedding` means that the model itself is a fusion model, but forgot that `FusionEmbedding` can be the embedding layer for a `TransformerDecoder`. In that case I agree, we should make a slight change here. Will leave a diff of what I think makes sense here

File: torchtune/modules/common_utils.py [hunk 9]
```diff
@@ -442,3 +448,131 @@ def delete_kv_caches(model: nn.Module):
...
+    """
+    need_resize = _need_to_resize(model, num_embeddings)
+    if not need_resize:
+        _log.info("Embedding resize not needed, skipping")
+        return
+
+    if not isinstance(model, TransformerDecoder) and need_resize:
+        raise AssertionError(
+            f"Resize needed but model is of type {type(model)}, please pass in TransformerDecoder to function"
+        )
```
Comment (ebsmothers) [inline torchtune/modules/common_utils.py:493]: Something like this? @pbontrager @iamzainhuda 

```suggestion
    if isinstance(model, DeepFusionModel) or isinstance(model, EarlyFusionModel):
        model = model.decoder
    if isinstance(model.tok_embeddings, FusionEmbedding) and resize_needed:
        raise AssertionError(
            f"Resize is not supported for FusionEmbeddings"
        )
```

File: torchtune/modules/common_utils.py (hunk 8 above)
Comment (iamzainhuda) [reply torchtune/modules/common_utils.py:453]: I've added `FusionEmbedding` resizing into the follow up tasks. I agree from looking at the module, we can extend the function to resize the fusion_embedding instead with ease
